packages/grid/src/vaadin-grid-data-provider-mixin.js

```diff
@@ -325,12 +325,22 @@ export const DataProviderMixin = (superClass) =>
 
     /** @protected */
     _onDataProviderPageReceived() {
+      const flatSizeChanged = this._dataProviderController.flatSize !== this._flatSize;
       // With the new items added, update the cache size and the grid's effective size
       this._flatSize = this._dataProviderController.flatSize;
 
       // After updating the cache, check if some of the expanded items should have sub-caches loaded
+      // or if some of the rows are missing an item
       this._getRenderedRows().forEach((row) => {
         this._dataProviderController.ensureFlatIndexHierarchy(row.index);
+        if (flatSizeChanged) {
```

Comment:
I guess it might be more efficient to do this work in `_debouncerApplyCachedData`, as changes in the hierarchy (if they happen shortly) could cause some items to get out of the viewport.

Reply:
IIRC it was done here since data provider controller hasn't yet [cleared](https://github.com/vaadin/web-components/blob/3633c1a76adc81c3fe1e9923150fe79ece9fcda3/packages/component-base/src/data-provider-controller/data-provider-controller.js#L256) the related pending request at this phase.

Reply:
Do you remember why it was important to do this before the pending request is removed from the queue?

Reply:
Possibly to avoid an [infinite update loop](https://github.com/vaadin/web-components/pull/7101/files#r1463393332)

Reply:
Moving this logic to the debouncer seems to fix the [issue](https://github.com/vaadin/web-components/pull/7101/files#r1464506154) reported by Sacha, in addition to fixing the original issue. However, this change also causes some test failures.